### ISSUE TITLE
faiss make error

### ISSUE THREAD

**Issue by 634671436** [facebookresearch/faiss#1314]
# Summary

when I  installed OpenBLAS and lapack,  then I git clone the faiss, cp the suitable linux makefile.inc, when I prepare to **make && make install** ,  I meet the IndexHnsw.h error, the detail is following:

faiss/IndexHNSW.cpp:429:0: warning: ignoring #pragma omp declare [-Wunknown-pragmas]
 #pragma omp declare reduction (stats_combine : HNSWStats : omp_out.combine(omp_in))
 ^
faiss/IndexHNSW.cpp:1047:0: warning: ignoring #pragma omp declare [-Wunknown-pragmas]
 #pragma omp declare reduction (stats_combine : HNSWStats : omp_out.combine(omp_in))
 ^
faiss/IndexHNSW.cpp: In member function 'virtual void faiss::IndexHNSW::search(faiss::Index::idx_t, const float*, faiss::Index::idx_t, float*, faiss::Index::idx_t*) const':
faiss/IndexHNSW.cpp:308:28: error: expected '+', '*', '-', '&', '^', '|', '&&', '||', 'min' or 'max' before 'stats_combine'
 #pragma omp for reduction (stats_combine : stats)
                            ^
faiss/IndexHNSW.cpp: In member function 'void faiss::IndexHNSW::search_level_0(faiss::Index::idx_t, const float*, faiss::Index::idx_t, const storage_idx_t*, const float*, float*, faiss::Index::idx_t*, int, int) const':
faiss/IndexHNSW.cpp:430:28: error: expected '+', '*', '-', '&', '^', '|', '&&', '||', 'min' or 'max' before 'stats_combine'
 #pragma omp for reduction (stats_combine : stats)
                            ^
faiss/IndexHNSW.cpp: In member function 'virtual void faiss::IndexHNSW2Level::search(faiss::Index::idx_t, const float*, faiss::Index::idx_t, float*, faiss::Index::idx_t*) const':
faiss/IndexHNSW.cpp:1048:28: error: expected '+', '*', '-', '&', '^', '|', '&&', '||', 'min' or 'max' before 'stats_combine'
 #pragma omp for reduction (stats_combine : stats)
                            ^
make: *** [faiss/IndexHNSW.o] Error 1

thanks for anybody to help!

**Comment by 634671436** [comment]
CXX          = g++ -std=gnu++0x -std=c++11
CXXCPP       = g++ -std=gnu++0x -std=c++11 -E
CPPFLAGS     = -DFINTEGER=int  -fopenmp -I/usr/local/cuda/include 
CXXFLAGS     = -fPIC -m64 -Wno-sign-compare -g -O3 -Wall -Wextra
CPUFLAGS     = -mpopcnt -msse4
LDFLAGS      = -fopenmp  -L/usr/local/cuda/lib64 
LIBS         = -lopenblas   -lcudart -lcublas 
PYTHONCFLAGS =  -I/search/odin/wangkun/anaconda3/include/python3.7m -I/search/odin/wangkun/anaconda3/include/python3.7m -I/search/odin/wangkun/anaconda3/lib/python3.7/site-packages/numpy/core/include
SWIGFLAGS    = -DSWIGWORDSIZE64

**Comment by 634671436** [comment]
CXX          = g++ -std=gnu++0x -std=c++11
CXXCPP       = g++ -std=gnu++0x -std=c++11 -E
CPPFLAGS     = -DFINTEGER=int  -fopenmp -I/usr/local/cuda/include 
CXXFLAGS     = -fPIC -m64 -Wno-sign-compare -g -O3 -Wall -Wextra
CPUFLAGS     = -mpopcnt -msse4
LDFLAGS      = -fopenmp  -L/usr/local/cuda/lib64 
LIBS         = -lopenblas   -lcudart -lcubla

**Comment by ChenChris613** [comment]
I encountered this situation too. Does anyone resolve this problem?

**Comment by mdouze** [comment]
Yes currently Faiss requires openmp 4 to compile. It is a bit worrying, I will see if it is possible to lift this requirement.

**Comment by macanv** [comment]
@mdouze how should we solve this problem？

**Comment by beauby** [comment]
A fix for this will be issued soon.

**Comment by mdouze** [comment]
fixed in internal FB version.

**Comment by sunicy** [comment]
> fixed in internal FB version.

There is another disturbing issue around for legacy compilers, say, GCC 4.8.x. With **--disable-openmp** set, the compiled static/shared object still looks for a bunch of _omp_* functions, including

0000000000000000         *UND*  0000000000000000 omp_destroy_lock
0000000000000000         *UND*  0000000000000000 omp_init_lock
0000000000000000         *UND*  0000000000000000 omp_get_max_threads
0000000000000000         *UND*  0000000000000000 omp_get_nested
0000000000000000         *UND*  0000000000000000 omp_set_nested
0000000000000000         *UND*  0000000000000000 omp_set_num_threads
0000000000000000         *UND*  0000000000000000 omp_in_parallel
0000000000000000         *UND*  0000000000000000 omp_get_num_threads
0000000000000000         *UND*  0000000000000000 omp_get_thread_num
...

OS: CentOS 6.8
GCC: 4.8.5
Faiss version: 1.6.3
Running on: GPU
Inferface: Python

Thanks for any update!

**Comment by mdouze** [comment]
Yes that is normal. There should be a stub omp implementation that says that there is a single thread, etc.

**Comment by sunicy** [comment]
> # Summary
> when I installed OpenBLAS and lapack, then I git clone the faiss, cp the suitable linux makefile.inc, when I prepare to **make && make install** , I meet the IndexHnsw.h error, the detail is following:
> 
> faiss/IndexHNSW.cpp:429:0: warning: ignoring #pragma omp declare [-Wunknown-pragmas]
> #pragma omp declare reduction (stats_combine : HNSWStats : omp_out.combine(omp_in))
> ^
> faiss/IndexHNSW.cpp:1047:0: warning: ignoring #pragma omp declare [-Wunknown-pragmas]
> #pragma omp declare reduction (stats_combine : HNSWStats : omp_out.combine(omp_in))
> ^
> faiss/IndexHNSW.cpp: In member function 'virtual void faiss::IndexHNSW::search(faiss::Index::idx_t, const float*, faiss::Index::idx_t, float*, faiss::Index::idx_t*) const':
> faiss/IndexHNSW.cpp:308:28: error: expected '+', '_', '-', '&', '^', '|', '&&', '||', 'min' or 'max' before 'stats_combine' #pragma omp for reduction (stats_combine : stats) ^ faiss/IndexHNSW.cpp: In member function 'void faiss::IndexHNSW::search_level_0(faiss::Index::idx_t, const float_, faiss::Index::idx_t, const storage_idx_t*, const float*, float*, faiss::Index::idx_t*, int, int) const':
> faiss/IndexHNSW.cpp:430:28: error: expected '+', '_', '-', '&', '^', '|', '&&', '||', 'min' or 'max' before 'stats_combine' #pragma omp for reduction (stats_combine : stats) ^ faiss/IndexHNSW.cpp: In member function 'virtual void faiss::IndexHNSW2Level::search(faiss::Index::idx_t, const float_, faiss::Index::idx_t, float*, faiss::Index::idx_t*) const':
> faiss/IndexHNSW.cpp:1048:28: error: expected '+', '*', '-', '&', '^', '|', '&&', '||', 'min' or 'max' before 'stats_combine'
> #pragma omp for reduction (stats_combine : stats)
> ^
> make: *** [faiss/IndexHNSW.o] Error 1
> 
> thanks for anybody to help!



> Yes that is normal. There should be a stub omp implementation that says that there is a single thread, etc.

Thanks for the reply. However, with those unresolved omp-related symbols, I failed to **import faiss** in Python 3.7. It either complains about unresolved symbols (but not those in omp, but in Cython, e.g., PyCObject_Type), or raises segmentation fault immediately. By enabling openmp and commenting #pragma_s_ where custom omp reductions declare, I could have Faiss built without any issue.

Thanks again for the reply!

**Comment by mdouze** [comment]
No activity, closing.
